working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+  MyRectangle affineTransform(int a, int b, int c, int d) => 
+      MyRectangle(a * left + b * top, c * left + d * top, 
+                  a * right + b * bottom, c * right + d * bottom);
+}
+```
+
+Here you implicitly "inherit" all constructors from the superclass without having to rewrite them.
+
+## Notes
+
+All the field initializers are named parameters. We do not want to make them positional parameters for two reasons: Primarily because it means that we have to pick an order, and alphabetical order is arbitrary, while source order makes the code unstable in the face of refactoring (and any other order is even more arbitrary). Secondarily because it makes it even harder to consistently forward to superclass constructors which also have named parameters.
+
+In NNBD libraries, we do not require that *final* instance variables are always initialized, but we do require that *potentially non-nullable* variables are. We could require that final fields are initialized, but it's very reasonable to treat being *nullable* as meaning being optional. Given that there is an initializing formal parameter for the field, it will definitely be initialized
+
+An instance variable in a subclass will always shadow a named parameter of the same name in a superclass constructor. That's why we do not forward to superclass constructors which have a required named parameter with such a name.
+
+Superclass constructors with optional positional parameters are hard to forward to. It would be wonderful if it was possible to have both optional positional parameters and named parameters on the same function.
+
+Existing mixin-application class forwarding constructors are just a special case of the general constructor forwarding functionality.
+
+We make constructors `const` only when the subclass declares no instance variables. We could allow final instance variables with no initializer expression (because then the constructor must initialize it) or with an initializer expression which is constant (because then it's valid in a `const` constructor class). The latter requirement is particularly fragile and useless. Changing the initializer expression of a variable from `final int x = a + 2;` to `final int x = a + b;` where `a` is a constant variable and  `b` is a non-constant variable with the value `2` will change the constructors of the class from constant to non-constant. Also, a final instance variable initialized to a constant value should probably just be a getter. We want to avoid accidentally making a class have constant constructors, because that may become a support burden for the author, so we only make constructors constant when the subclass is particularly simple, and where the current mixin application constructor forwarding makes the constructor constant. (It is not a *breaking* change to make more constructors constant in the future, but it is a *bad* change for library owners which intended a class to have non-constant constructors).
```

Comment:
+1, though see my above suggestion about maybe letting users opt in to const default constructors.

Reply:
Also reasonable when you can write `const default Foo;` and not have to write all the parameters.